system/trace/trace_dump.c

```diff
@@ -0,0 +1,643 @@
+#if CONFIG_TASK_NAME_SIZE > 0
+          FAR struct note_start_s *nst = (FAR struct note_start_s *)p;
+          FAR struct trace_dump_task_context_s *tctx;
+
+          tctx = get_task_context(pid, ctx);
+          if (tctx != NULL)
+            {
+              copy_task_name(tctx->name, nst->nst_name);
+            }
+#endif
+
+          trace_dump_header(out, note, ctx);
+          fprintf(out, "sched_wakeup_new: comm=%s pid=%d target_cpu=%d\n",
+                  get_task_name(pid, ctx), get_pid(pid), cpu);
+        }
+        break;
+
+      case NOTE_STOP:
+        {
+          trace_dump_header(out, note, ctx);
+          fprintf(out, "sched_switch: "
+                       "prev_comm=%s prev_pid=%u prev_state=%c ==> "
+                       "next_comm=%s next_pid=%u\n",
+                  get_task_name(pid, ctx), get_pid(pid), 'X',
+                  get_task_name(cctx->current_pid, ctx),
+                  get_pid(cctx->current_pid));
+        }
+        break;
+
+      case NOTE_SUSPEND:
+        {
+          FAR struct note_suspend_s *nsu = (FAR struct note_suspend_s *)p;
+
+          /* This note informs the task to be suspended.
+           * Preserve the information for the succeeding NOTE_RESUME.
+           */
+
+          cctx->current_state = nsu->nsu_state;
+        }
+        break;
+
+      case NOTE_RESUME:
+        {
+          /* This note informs the task to be resumed.
+           * The task switch timing depends on the running context.
+           */
+
+          cctx->next_pid = pid;
+
+          if (!cctx->ininterrupt)
+            {
+              /* If not in the interrupt context, the task switch is
+               * executed immediately.
+               */
+
+              trace_dump_header(out, note, ctx);
+              trace_dump_sched_switch(out, note, ctx);
+            }
+          else
+            {
+              /* If in the interrupt context, the task switch is postponed
+               * until leaving the interrupt handler.
+               */
+
+              trace_dump_header(out, note, ctx);
+              fprintf(out, "sched_waking: comm=%s pid=%d target_cpu=%d\n",
+                      get_task_name(cctx->next_pid, ctx),
+                      get_pid(cctx->next_pid), cpu);
+              cctx->pendingswitch = true;
+            }
+        }
+        break;
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+      case NOTE_SYSCALL_ENTER:
+        {
+          FAR struct note_syscall_enter_s *nsc;
+          FAR struct trace_dump_task_context_s *tctx;
+
+          /* Exclude the case of syscall issued by an interrupt handler and
+           * nested syscalls to correct tracecompass display.
+           */
+
+          if (cctx->ininterrupt)
+            {
+              break;
+            }
+
+          tctx = get_task_context(pid, ctx);
+          if (tctx == NULL)
+            {
+              break;
+            }
+
+          tctx->syscall_nest++;
+          if (tctx->syscall_nest > 1)
```

Comment:
The syscall nest while executing the interrupt handler should not belong to the syscall nest in the task.
For now, because all syscalls in the interrupt handler cannot be displayed, we can ignore the syscall immediately if in the interrupt handler.